hypervisor/src/arch/x86/emulator/instructions/mod.rs

```diff
@@ -0,0 +1,87 @@
+//
+// Copyright © 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+
+extern crate iced_x86;
+
+use crate::arch::emulator::EmulatorError;
+use crate::arch::x86::emulator::instructions::mov::Mov;
+use crate::arch::x86::emulator::IOEmulator;
+use iced_x86::*;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+fn operand_memory_address(
+    insn: &Instruction,
+    io: &Arc<Mutex<dyn IOEmulator>>,
+) -> Result<u64, EmulatorError> {
+    let mut address: u64 = 0;
+    let io = io.lock().unwrap();
+
+    if insn.memory_base() != iced_x86::Register::None {
+        let base: u64 = io.read_reg(insn.memory_base())?;
+        address += base;
+    }
+
+    if insn.memory_index() != iced_x86::Register::None {
+        let mut index: u64 = io.read_reg(insn.memory_index())?;
+        index *= insn.memory_index_scale() as u64;
+
+        address += index;
+    }
+
+    address += insn.memory_displacement() as u64;
+
+    Ok(address)
+}
+
+fn operand_memory_size(insn: &Instruction) -> Result<usize, EmulatorError> {
+    let size = match insn.memory_size() {
+        MemorySize::UInt8 => Ok(1),
+        MemorySize::UInt16 => Ok(2),
+        MemorySize::UInt32 => Ok(4),
+        MemorySize::UInt64 => Ok(8),
+        _ => Err(EmulatorError::UnsupportedMemorySize(anyhow!(
+            "{:?}",
+            insn.memory_size()
+        ))),
+    }?;
+
+    Ok(size)
+}
+
+pub trait InstructionHandler {
+    fn emulate(
+        &self,
+        insn: &Instruction,
+        io: Arc<Mutex<dyn IOEmulator>>,
+    ) -> Result<(), EmulatorError>;
+}
+
+pub struct InstructionMap {
+    pub instructions: HashMap<Mnemonic, Box<dyn InstructionHandler + Sync + Send>>,
```

Comment:
Would it be better to map from Code to instruction handler instead of Mnemonic? It seems like this approach throws away and rederives information that iced_x86 already provides in the form of the code.

E.g. The Mnemonic::Mov maps from 40 different instructions with different widths and operand types. You have to re-derive this information in your mov instruction handler.

Reply:
I was hesitant to do so when I originally wrote that code but I took your advice and gave it a try. The code is more compact and can be factorized through a few macros. Please have a look.